ERCS/erc-7812.md

```diff
@@ -0,0 +1,514 @@
+
+    /**
+     * @notice Gets the element value by its key.
+     */
+    function getValue(bytes32 key) external view returns (bytes32);
+}
+```
+
+### Evidence Registry
+
+The `EvidenceRegistry` smart contract is the central piece of this EIP. The `EvidenceRegistry` MUST implement the following interface, however, it MAY be extended:
+
+```solidity
+pragma solidity ^0.8.0;
+
+/**
+ * @notice Common Evidence Registry interface.
+ */
+interface IEvidenceRegistry {
+    /**
+     * @notice MUST be emitted whenever the Merkle root is updated.
+     */
+    event RootUpdated(bytes32 indexed prev, bytes32 indexed curr);
+
+    /**
+     * @notice Adds the new statement to the DB.
+     */
+    function addStatement(bytes32 key, bytes32 value) external;
+
+    /**
+     * @notice Removes the statement from the DB.
+     */
+    function removeStatement(bytes32 key) external;
+
+    /**
+     * @notice Updates the statement in the DB.
+     */
+    function updateStatement(bytes32 key, bytes32 newValue) external;
+
+    /**
+     * @notice Retrieves historical DB roots creation timestamps.
+     * Latest root MUST return `block.timestamp`.
+     * Non-existent root MUST return `0`.
+     */
+    function getRootTimestamp(bytes32 root) external view returns (uint256);
+}
+```
+
+The `addStatement`, `removeStatement`, and `updateStatement` methods MUST isolate the statement `key` in order for the database to allocate a specific namespace for a caller. These methods MUST revert in case the isolated key being added already exists in the `EvidenceDB` or the isolated key being removed or updated does not.
+
+The `EvidenceRegistry` MUST maintain the linear history of `EvidenceDB` roots. The `getRootTimestamp` method MUST NOT revert. Instead, it MUST return `0` in case the queried `root` does not exist. The method MUST return `block.timestamp` in case the latest root is requested.
+
+Before communicating with the `EvidenceDB`, the `key` MUST be isolated in the following way:
+
+```solidity
+bytes32 isolatedKey = hash(msg.sender, key)
+```
+
+Where the `hash` is secure protocol-wide hash function of choice.
+
+### Hash Function
+
+The same secure hash function MUST be employed in both `EvidenceRegistry` and `EvidenceDB`. It is RECOMMENDED to use ZK-friendly hash function such as `poseidon` to streamline the database proving.
+
+In case ZK-friendly hash function is chosen, `EvidenceRegistry` MUST NOT accept `keys` or `values` beyond the underlying elliptic curve prime field size (`21888242871839275222246405745257275088548364400416034343698204186575808495617` for `BN128`). 
+
+## Rationale
+
+This EIP stemmed from the need to localize and unravel the storage and issuance of provable statements so that future protocols can anchor to the standardized singleton registry. The proposal is deliberately written as abstract as possible to not constraint the possible business use cases and allow `Registrars` to implement arbitrary provable solutions.
```

Comment:
> This EIP stemmed from the need to [...]

Content like this belongs in the motivation section, because it's describing the document as a whole. Motivation should justify the proposal as a whole, while the rationale should explain choices made within the document itself. My pet analogy is:

> Motivation: We need to build a shed because...
> Rationale: We chose to paint the shed red because...